api/base_controller.py

```diff
@@ -85,9 +75,21 @@ def authenticated_patron(self, authorization_header):
 
     def authenticate(self):
         """Sends a 401 response that demands authentication."""
-        if not self.manager.opds_authentication_document:
-            self.manager.opds_authentication_document = self.manager.auth.create_authentication_document()
+        library_short_name = flask.request.library.short_name
+        if library_short_name not in self.manager.opds_authentication_documents:
+            self.manager.opds_authentication_documents[library_short_name] = self.manager.auth.create_authentication_document()
 
-        data = self.manager.opds_authentication_document
+        data = self.manager.opds_authentication_documents[library_short_name]
         headers = self.manager.auth.create_authentication_headers()
         return Response(data, 401, headers)
+
+    def library_from_request(self, library_short_name):
```

Comment:
I agree it's not an accurate name but since it modifies `flask.request` I like keeping 'request' in the name. Maybe library_for_request or add_library_to_request?